bpo-27099: IDLE - Convert built-in extensions to regular features

wohlganger opened the pull request:
* Adds fg color, bg color, font, and underline options for parens highlighting

<!-- issue-number: bpo-27099 -->
https://bugs.python.org/issue27099
<!-- /issue-number -->

wohlganger committed:
parenmatch highlighting options

adds fg color, bg color, font, and underline options for parens highlighting

wohlganger committed:
fixed whitespace issue

wohlganger committed:
only select from highlighting options

wohlganger committed:
fixed whitespace

wohlganger commented:
Per Terry J. Reedy recommendation, only highlighting style is changeable.

wohlganger committed:
whitespace

wohlganger committed:
Merge remote-tracking branch 'upstream/master'

wohlganger committed:
#27099 - turn builtin extensions to reg modules

wohlganger committed:
whitespace + fix I thought I already committed.

wohlganger committed:
warning fixes, removed old config entries

wohlganger committed:
calltip append doc fix,  default paren-fore fix, help update

How did the change to the __doc__ of append not get fixed in the tests earlier?

wohlganger committed:
fix calltips

Used incorrect / old python for testing.

wohlganger committed:
removed circular import

wohlganger committed:
fix warnings due to missing default config options in idlelib.config-main.def

wohlganger committed:
Merge branch 'master' into master

wohlganger committed:
added format_pmaxw to config-main

wohlganger committed:
Merge remote-tracking branch 'upstream/master'

terryjreedy commented on Lib/idlelib/help.html:
```diff
@@ -636,19 +636,8 @@
-<li>RstripExtension</li>
-</ul>
+in the idlelib directory for further information.  There are currently no default extensions.</p>
+
```
help.html is generated from Doc/library/idle.rst.  So replace this edit with an edit of the latter.  I believe it should be straightforward.

wohlganger committed:
replaced help.html edit with idle.rst edit

per terryjreedy request.

wohlganger committed:
Merge branch 'master' into master

terryjreedy commented:
Charles, I cannot add commits to this issue, and a few others. Please follow the instructions at https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/#enabling-repository-maintainer-permissions-on-existing-pull-requests to make sure "Allow edits from maintainers." is checked for this and other IDLE issues. Please let me know what you find. If this permission is already granted, then I have some other problem with github.

wohlganger commented:
Checked "Allow edits". Thank you for pointing it out; I'll remember to check it in the future.

terryjreedy commented:
It is on now.  'Allow edits' is supposed to be checked by default.  If that changed, let me know and I will pass the information on to whoever can change it back.

wohlganger committed:
Merge remote-tracking branch 'upstream/master'

wohlganger committed:
added conflict resolved var descriptions to docstring in create_page_highlight.

wohlganger committed:
strip trailing whitespace

wohlganger committed:
Merge branch 'master' into master

wohlganger committed:
re-introduced highlight options.

Fixing code conflict ate highlighting options. Needed to be re-introduced.

terryjreedy committed:
News blurb.

terryjreedy commented:
Because of the updates, this merged cleanly, so I added the news file.

I will worry about the IdleConfTest failures after looking at changes to other files.  Appeveyor also had this.

FAIL: test_keybinding (idlelib.idle_test.test_configdialog.KeysPageTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\cpython\lib\idlelib\idle_test\test_configdialog.py", line 838, in test_keybinding
    {'AutoExpand_cfgBindings': {'expand-word': '<Key-F11>'}})
AssertionError: {} != {'AutoExpand_cfgBindings': {'expand-word': '<Key-F11>'}}
- {}
+ {'AutoExpand_cfgBindings': {'expand-word': '<Key-F11>'}}

wohlganger commented:
The configdialog test is failing because it's expecting AutoExpand to be an extension, which it is not. In order for the test to be meaningful, there needs to be a default extension. I'm going to make ZoomHeight that default extension; it seems to be the ideal candidate.

wohlganger committed:
Reinstated zoomheight as an extension. Replaced autoexpand with zoomheight as  non-core keybinding test example.

wohlganger committed:
fixed configdialog test (wrong cases), fixed whitespace

wohlganger committed:
update idle.rst: zoomheight is an extension.

wohlganger committed:
more fixes for zoomheight - trying to minimize changes from master

wohlganger committed:
zoomheight config fix, zoomheight should be fully back to the way it was before.

wohlganger committed:
whitespace

wohlganger committed:
fixed force-open completions missing from config-keys and config. Fixed zoomheight configdialog test.

terryjreedy commented on Lib/idlelib/configdialog.py:
```diff
@@ -208,12 +207,16 @@ def deactivate_current_config(self):
 
+
+
+
```
Remove this addition and leave single spaced. 
Apparently done.

terryjreedy commented on Lib/idlelib/configdialog.py:
```diff
@@ -208,12 +207,16 @@ def deactivate_current_config(self):
         Dynamically update the current parent window instances
         with some of the configuration changes.
         """
+
```
Remove. -- done

terryjreedy commented on Lib/idlelib/configdialog.py:
```diff
@@ -1099,6 +1169,7 @@ def create_new(self, new_theme_name):
             save_new
             set_theme_type
         """
+
```
Remove.  Please don't go around reformatting, especially is a way that counters existing consistent style. -- apparently done here.

terryjreedy commented on Lib/idlelib/configdialog.py:
```diff
@@ -2008,33 +2079,7 @@ def detach(self):
-'opener' - opener '({[' corresponding to closer; 'parens' - both chars;
-'expression' (default) - also everything in between.  Flash-delay is how
-long to highlight if cursor is not moved (0 means forever).
-'''
```
Revert this deletion.  'Extensions' key should become 'General' after widgets are moved there.
Reversion done. Name change later.

terryjreedy reviewed (changes requested):
configdialog.py

wohlganger committed:
Changes made per terryjreedy request.

Sorry about the formatting; I think that happened when I was trying to fix for the conflicts.

wohlganger commented:
please wait before checking - I haven't fixed the entry for autocomplete and paragraph yet.

wohlganger committed:
re-added autocomplete wait and format paragraph max width options to general tab

wohlganger commented:
Ready to be checked

wohlganger commented:
looking at the four failing tests, it wouldn't be difficult to fix them so that they pass with this patch, but the easy solution assumes there will always only be one extension - ZoomHeight. Future proofing it would be better, but I think that's outside the scope of this patch. Do you want me to fix those tests?

terryjreedy commented on Lib/idlelib/autocomplete.py:
```diff
@@ -27,18 +26,10 @@
-
-    popupwait = idleConf.GetOption("extensions", "AutoComplete",
-                                   "popupwait", type="int", default=0)
-
```
Why did you move this?  Although another issue, really, I can see moving within __init__ to have changes take effect  sooner, why make it conditional?

terryjreedy commented on Lib/idlelib/autocomplete.py:
```diff
@@ -47,6 +38,10 @@ def __init__(self, editwin=None):
+    def reset(self):
+        self.popupwait = idleConf.GetOption("main", "General",
+                                   "autocomplete_wait", type="int", default=0)
+
```
One-liner is not worth a separate function.

wohlganger replied:
Built-in options are effective when they are changed from the config menu. Extensions require restarting IDLE. In order for these to function like built-ins, their changed settings needed to be reloaded from within the editor. That is what the reset function is for. There is one in each of the old extensions that had options. It seems the activation got lost when I was resolving conflicts, but it looks like the place to call it is now activate_config_changes in the configdialog. Would you rather I put the functionality back in, or that some of the new options behave differently than the old ones?

terryjreedy replied:
I opened bpo-22707 two years ago for this problem, and intended that we work on it after this one. I just posted my analysis and made you nosy.  For reasons explained there, I want to leave the class attributes as class attributes.  Whether the idleConf wrapper are added here or in 22707, they should be called 'reload'; 'reset' implies changing back to initial values.  They should be classmethods called after the class statement.  This will preserve the current behavior, which I will manually test.

 The line length for ReformatParagraph should be a class attribute, and the test adjusted.  CodeContext().visible should be left alone until fixed or deleted.

I want to deal with changing code to update immediately in 22707, where the new code will be tested.  I will add more on that issue.

terryjreedy commented on Lib/idlelib/autocomplete.py:
```diff
@@ -1,6 +1,5 @@
+"""
+This can complete either attribute names or file names. It can pop
 a window with all available names, for the user to select from.
 """
```
"""Complete attribute name or file name.

Either on demand or after a user-selected delay after a key character, pop up a list of candidates.
"""
Note no return after initial quote.

terryjreedy commented on Lib/idlelib/autoexpand.py:
```diff
@@ -10,7 +10,6 @@
 place before requesting the next selection causes AutoExpand to reset
 its state.
 
-This is an extension file and there is only one instance of AutoExpand.
```
Leave 'There is only one instance of AutoExpand'.
This docstring actually belongs with the class, and will have to be moved when files are consolidated -- in another issue.

terryjreedy commented on Lib/idlelib/codecontext.py:
```diff
@@ -45,15 +39,28 @@ def __init__(self, editwin):
+        if self.topvisible:
+            #need to rebuild widget to change color
+            self.toggle_code_context_event()
+            self.toggle_code_context_event()
```
This and other edits is an attempt to change the behavior of the feature, as opposed to simply changing the internal implementation.  These behavior changes have not been discussed, and mostly should be one or more separate issues.  Some of theme may already be so.

terryjreedy commented on Lib/idlelib/config-extensions.def:
```diff
@@ -32,67 +32,6 @@
 
-[AutoComplete]
-enable=True
-popupwait=2000
```
As I said and explained on the issue, I want to leave non-key options in config-extensions.  So the 4 non-trivial [Xtension] sections should stay.  For former extensions, I would like to drop the requirement for an enable option when the entry is read in the editor.  Then we can delete the enable items.  I believe that this will be back-compatible with existing custom enable settings, but must verify.    (Ditto for not requiring some binding section.)

As I said yesterday, I now think the [Xtension_cfgBindings] should go, to be replaced by new entries in keysets.  This will require some change in IdleConf to deal better with missing entries in custom keysets.   But this was already needed, and perhaps should be a separate preliminary patch.

I think putting non-configurable key bindings in configuration files was a bit of a mistake.  In any case, these binding should be hard coded in the appropriate place.  At minimum, 'x_keys = idleConf(...)' should be replaced by 'x_keys = (current-idleconf-return).

terryjreedy commented on Lib/idlelib/config-extensions.def:
```diff
@@ -32,67 +32,6 @@
-numlines=3
-visible=False
-bgcolor=LightGray
-fgcolor=Black
```
Keep

terryjreedy commented on Lib/idlelib/config-extensions.def:
```diff
@@ -32,67 +32,6 @@
-
-[FormatParagraph]
-enable=True
-max-width=72
```
Keep

terryjreedy commented on Lib/idlelib/config-extensions.def:
```diff
@@ -32,67 +32,6 @@
-style= expression
-flash-delay= 500
-bell=True
-[ParenMatch_cfgBindings]
```
Keep.

terryjreedy commented on Lib/idlelib/config-highlight.def:
```diff
@@ -22,6 +22,10 @@ hit-foreground= #ffffff
+codecontext-foreground = black
+codecontext-background = LightGray
+parenmatch-foreground = black
+parenmatch-background = gray
```
Delete parenmatch additions.

terryjreedy commented on Lib/idlelib/config-keys.def:
```diff
@@ -57,6 +57,19 @@ toggle-tabs=<Alt-Key-t> <Meta-Key-t> <Alt-Key-T> <Meta-Key-T>
+flash-paren=<Control-Key-0>
+paren-closed=<KeyRelease-parenright> <KeyRelease-bracketright> <KeyRelease-braceright>
+run-module=<Key-F5>
+check-module=<Alt-Key-x>
```
Only current user-configurable 'event = keylist' pairs from [Xtension-cfgBindings] sections should be added.  These are
```
force-open-completions= <Control-Key-space>
expand-word= <Alt-Key-slash>
force-open-calltip= <Control-Key-backslash>
format-paragraph= <Alt-Key-q>
flash-paren= <Control-Key-0>
run-module= <Key-F5>
check-module= <Alt-Key-x>
zoom-height= <Alt-Key-2>
```
The keys for autocomplete, try-open-completions, try-open-calltip, refresh-calltip, and paren-closed are not and should not be configurable.  Adding new shortcuts, such as for toggle-code-context, will be a separate issue.

terryjreedy commented on Lib/idlelib/config-main.def:
```diff
@@ -74,6 +76,11 @@ num-spaces= 4
+numlines= 3
+parenstyle= opener
+flash-delay= 500
+bell= 1
```
Delete additions

terryjreedy commented on Lib/idlelib/config.py:
```diff
@@ -359,7 +359,12 @@ def GetThemeDict(self, type, themeName):
+                'codecontext-foreground':'#000000',
+                'codecontext-background':'#ffffff',
+                'parenmatch-foreground':'#000000',
+                'parenmatch-background':'#ffffff',
```
Delete parenmatch.

terryjreedy commented on Lib/idlelib/config.py:
```diff
@@ -654,7 +659,21 @@ def GetCoreKeys(self, keySetName=None):
+            '<<flash-paren>>':['<Control-Key-0>'],
+            '<<paren-closed>>':['<KeyRelease-parenright>', '<KeyRelease-bracketright>', '<KeyRelease-braceright>'],
+            '<<run-module>>':['<Key-F5>'],
+            '<<check-module>>':['<Alt-Key-x>']
```
Reduce additions to current user-configurable bindings.

terryjreedy commented on Lib/idlelib/configdialog.py:
```diff
@@ -770,6 +769,8 @@ def create_page_highlight(self):
             'Shell Stdout Text': ('stdout', '12'),
             'Shell Stderr Text': ('stderr', '13'),
+            'Code Context Text': ('codecontext', '14'),
+            'Matched Parenthetics': ('parenmatch', '15'),
```
delete

terryjreedy commented on Lib/idlelib/configdialog.py:
```diff
@@ -782,26 +783,48 @@ def create_page_highlight(self):
+        self.flash_delay = tracers.add(
+                StringVar(self), self.var_changed_flash_delay)
+        self.num_lines = tracers.add(
+                StringVar(self), self.var_changed_num_lines)
```
Move to General tab, though possible after I condense current stuff.

terryjreedy commented on Lib/idlelib/configdialog.py:
```diff
@@ -782,26 +783,48 @@ def create_page_highlight(self):
-            ('func', 'definition'), ('(param):\n  ', 'normal'),
-            ('"""string"""', 'string'), ('\n  var0 = ', 'normal'),
+            ('func', 'definition'),
+            ('(parenthetics)','parenmatch'),
```
delete

terryjreedy commented on Lib/idlelib/editor.py:
```diff
@@ -57,6 +58,15 @@ class EditorWindow(object):
+        from idlelib.paragraph import FormatParagraph
+        from idlelib.parenmatch import ParenMatch
+        from idlelib.rstrip import RstripExtension
+        from idlelib.runscript import ScriptBinding
```
Why here instead with other class imports?  Is there a circular import problem?

wohlganger replied:
Travis or AppVeyor would error out with them when they were with the other imports, so I was pretty sure it was a circular import.

terryjreedy replied:
I am sure that some of the imports, such as the one for paragraph, could be moved to class level.  They should be tested one by one, and a comment added for those that must stay.   I will eventually do this is you don't.  I hope to eventually reduce the loopy dependencies, so I want to know which fail otherwise.  I also want to make sure the errors are from circular imports and not something else that should perhaps be fixed.

terryjreedy replied:
To do import just once, on first instance: at class level, add for instance 'AutoComplete = None' and in __init__: 'if AutoComplete == None: self.do_imports()' where do_imports does delayed imports and sets class attributes.

terryjreedy commented on Lib/idlelib/editor.py:
```diff
@@ -270,6 +280,32 @@ def __init__(self, flist=None, filename=None, key=None, root=None):
         self.showerror = tkMessageBox.showerror
 
+
+        #init merged extentions binds - needs to be done after color is set
```
Can you explain comment?

wohlganger replied:
Code context and some of the other extensions depend on frame.text being packed, which occurs in color_config, called from ResetColorizer. If done before then, idle errors out.

terryjreedy commented on Lib/idlelib/editor.py:
```diff
@@ -270,6 +280,32 @@ def __init__(self, flist=None, filename=None, key=None, root=None):
 
+
+        #init merged extentions binds - needs to be done after color is set
+        self.insAutoComplete=AutoComplete(self)
```
Spaces around '=', as per PEP8.

terryjreedy reviewed (changes requested):
The overall problem is that the patch does more than specified on the issue. See the separate comments.

terryjreedy commented:
The tests must pass one way or another.  I have thought about a possible dummy extension entry and file that would properly illustrate the features, but leaving zoomheight as is for now seems ok.  It might even be left as an (optional) extension indefinitely.  We could add comments to make it serve better as didactic example.

terryjreedy commented:
After re-reading bpo-20580 and reading the sed commands in https://hg.python.org/cpython/rev/13826ff147e4, leaving zoomheight permanently an extension is not ok.  For reasons explained in the issue, Ned would like to get rid of the sed edits.

wohlganger committed:
Merge branch 'master' into master

wohlganger commented:
It's getting unclear where you want the options. If they are left in config-extensions.def, they are automatically populated in the extensions tab. If there's the ability to change them from the General tab, users would expect them to be visible in .ilderc\config-main.cfg . Should I move them to config-main.def, or is there something I'm missing? Additionally - if we need the tests to pass AND we're moving zoomheight out of extensions, which causes the tests to fail - what is the plan there?

terryjreedy commented:
I have consistently said that I want the widgets on the general tab.   I also want to leave then in config-extensions.def for back-compatibility. They will not appear on the extension tab. With 'enable' removed and with no binding, the truncated entries should not be recognized as legal extension entries.  We just need to make sure that they are ignored and no error or warning is emitted.  Sorry if I did not make this clear enough before.

Something similar is needed for the editor extension loading.  If nothing else, we can add an explicit exclusion list, but I don't expect to need that.  If you don't beat me to it, I will write both changes.

Users do not need to ever look at .idlerc/config-main.cfg.  It should only be edited with the dialog. We will add comments in the idlelib .def versions explaining the situation.

I plan on adding an annotated, functional, dummy extension, disabled by default.  It will be both an example of how to write an extension and a test subject.

wohlganger committed:
Merge branch 'master' into master

wohlganger committed:
changes per terryjreedy

still need to move the imports

wohlganger committed:
fix warnings, errors, bugs

AutoComplete, CodeContext, FormatParagraph, ParenMatch needed special exclusions to have settings read from config-extensions.def . Special exclusions needed for calltips, completions, paren-closed events to prevent them being configurable.

wohlganger committed:
whitespace

wohlganger committed:
bugfix

if the cfg file had one of the old extensions in it, it was showing up in the extensions tab. This fixes that.

wohlganger commented:
I think I've gotten this patch how you want it. Since there's no dummy extension for testing, it still throws loads of errors in travis and appveyor.

terryjreedy commented:
Good.  I will try to re-review in an hour or two.  I will add an extension and see what errors remain.

terryjreedy commented on Lib/idlelib/autocomplete.py:
```diff
@@ -27,25 +26,20 @@
-    def __init__(self, editwin=None):
+    def __init__(self, editwin):
         self.editwin = editwin
-        if editwin is not None:  # not in subprocess or test
```
Removing '=None' and the if statement caused subprocess startup to fail.
Fixed.

terryjreedy commented on Lib/idlelib/autocomplete.py:
```diff
@@ -27,25 +26,20 @@
+
+    def reload(self):
+        self.popupwait = idleConf.GetOption("main", "General",
+                                   "autocomplete_wait", type="int", default=0)
```
I want these to be class methods called after the class is created.  I will fix for all 4 files. I also changed back to original args -- 'extensions', 'AutoComplete', 'popupwait'.

terryjreedy commented on Lib/idlelib/config.py:
```diff
@@ -438,6 +439,10 @@ def GetExtensions(self, active_only=True,
+        for extn in ['AutoComplete','CodeContext','FormatParagraph','ParenMatch']:
+            extns.remove(extn)
+            # specific exclusions because we are storing config for mainlined old
+            # extensions in config-extensions.def for backward compatibility
```
Right code, wrong place.  For me, IDLE startup failed because I have customized AutoComplete.popupwait and therefore have an AutoComplete entry in config-extensions.cfg, which resulted in AutoComplete being added back to extns by the next 4 lines.  Problem fixed by moving exclusions after extns building is complete.

terryjreedy committed:
Make changes need to get IDLE to run

terryjreedy reviewed (commented):
I could not work on this last night, but I am now.  Please to not push any more commits.  I had to make a couple of changes to get IDLE to run.  I will make other changes too.  When I get tests passing, I will push everything.

terryjreedy committed:
Edit extension/feature files, making existing tests pass.

terryjreedy committed:
Redo feature bindings.

terryjreedy committed:
First draft of dummy extension.

terryjreedy committed:
Fix that eliminate startup error and all but 2 test_idle errors.

terryjreedy commented:
I am not done: there are 2 errors left in text_config, + I want to reshuffle widgets on General tab.
+ more gracefully handling 'missing' entries in custom keysets.

terryjreedy committed:
Move widgets, load widgets, add tests, make previous tests pass.

terryjreedy commented:
There were more that two errors.  Anyway, I think this is complete enough.  There is a unittest for each new widget, and all tests pass for me.  I though the context highlight had moved to the highlights tabs, but if so, that disappeared.  I am willing to leave that to separate code context patch.  A method of dealing with the missing highlight in the user highlight file will be needed.  Ditto for dealing with missing keys in user key file.  

I still need to do manual lives test all the transported features.

wohlganger commented:
I had moved codecontext and parenmatch options into the highlight tab and integrated them with the colorizer system, but you said you wanted them in general, so I moved them back. I think it'll make more sense to move them in a later patch than to make this any bigger than it needs to be, especially since it's passing tests now.

terryjreedy commented on Lib/idlelib/config-keys.def:
```diff
@@ -57,6 +57,14 @@ toggle-tabs=<Alt-Key-t> <Meta-Key-t> <Alt-Key-T> <Meta-Key-T>
+force-open-completions= <Control-Key-space>
+expand-word= <Alt-Key-slash>
+force-open-calltip= <Control-Key-backslash>
+format-paragraph= <Alt-Key-q>
```
On the Format menu, format-paragraph does not get its accelerator (hot-key).  In addition, `<<format-paragraph>>` does not appear in the ConfigDialog  keys list.  (The menu entry does work.)  The other 7 additions have their accelerators on the menu and event-seq pairs on the keys tab.

The reason was the absence of `<<format-paragraph>>` is from the core key list when I start with a custom key set.

terryjreedy commented on Lib/idlelib/config.py:
```diff
@@ -654,7 +659,17 @@ def GetCoreKeys(self, keySetName=None):
+            '<<paren-closed>>':['<KeyRelease-parenright>', '<KeyRelease-bracketright>', '<KeyRelease-braceright>'],
+            '<<run-module>>':['<Key-F5>'],
+            '<<check-module>>':['<Alt-Key-x>'],
+            '<<zoom-height>>':['<Alt-Key-2>']
```
The additions, starting with `<<try-open-calltip>>`, include 7 of 8 user-configurable events, missing `<<format-paragraph>>`.   This prevented the needed substitution when using a custom keyset.  Fixed.

The additions also include 3 of 6 fixed events, missing <<autocomplete>>, <<refresh-call-tips>>, and <<toggle-code-context>>.  The purpose, I finally understand, was to bring about event_add.  But adding fixed events to the list of configurable pairs results in them appearing on the configure Keys page.  I will add explicit statements elsewhere.

terryjreedy committed:
Add format-paragraph to core keys.

terryjreedy committed:
Merge remote-tracking branch 'origin/master' into pr_2494

terryjreedy commented on Lib/idlelib/config-extensions.def:
```diff
@@ -32,68 +52,14 @@
-[AutoComplete_cfgBindings]
-force-open-completions=<Control-Key-space>
-[AutoComplete_bindings]
-autocomplete=<Key-Tab>
```
The '_bindings' entries had two effects:
1. event_add a virtual event associated with the event sequences.
2. bind the virtual event to an event handler (named after the event).
This new line in EditorWindow did the second:
`        text.bind("<<autocomplete>>", autocomplete.autocomplete_event)`
The lack of a replacement for 1. disabled the use of tab to open and tab tab to close a completion box.  Adding
`            self.text.event_add('<<autocomplete>>', '<Key-Tab>')`
in AutoComplete.__init__ fixed the problem.  I am thinking that both parts should be in the same file.  Either the event_add should be in EditorWindow or the binding should be in AutoComplete.  I am leaning toward the latter, but will decide tomorrow.

I decided to put all event_adds in EditorWindow for now.  I believe they only need to be called once in the process, not once for every window, but must check and then decide where to put them.

terryjreedy commented on Lib/idlelib/config-extensions.def:
```diff
@@ -32,68 +52,14 @@
-force-open-completions=<Control-Key-space>
-[AutoComplete_bindings]
-autocomplete=<Key-Tab>
-try-open-completions=<KeyRelease-period> <KeyRelease-slash> <KeyRelease-backslash>
```
Here you accomplished effect 1 above, `event_add(...)`, by adding the event-sequences pair to core bindings.  But this has the bug of adding then to the configure Keys tab.  Core bindings are configurable bindings.  I will add `event_add` elsewhere when I decide where.

terryjreedy commented on Lib/idlelib/config-extensions.def:
```diff
@@ -32,68 +52,14 @@
-[CallTips_cfgBindings]
-force-open-calltip=<Control-Key-backslash>
-[CallTips_bindings]
-try-open-calltip=<KeyRelease-parenleft>
```
Ditto.

terryjreedy commented on Lib/idlelib/config-extensions.def:
```diff
@@ -32,68 +52,14 @@
-force-open-calltip=<Control-Key-backslash>
-[CallTips_bindings]
-try-open-calltip=<KeyRelease-parenleft>
-refresh-calltip=<KeyRelease-parenright> <KeyRelease-0>
```
This needs event_add.
The presence of <KeyRelease-0> was a bug that caused calltip to flash off and on.  I removed it.

terryjreedy commented on Lib/idlelib/config-extensions.def:
```diff
@@ -32,68 +52,14 @@
-bgcolor=LightGray
-fgcolor=Black
-[CodeContext_bindings]
-toggle-code-context=
```
The is a dummy line as there is not key to associate with a virtual event.

terryjreedy reviewed (changes requested):
The fixed key bindings are not yet handled properly, resulting in multiple bugs.  Except for one detail, I believe I know what to do and will do so tomorrow, and then retest.

terryjreedy committed:
Fix autocomplete and long lines.

terryjreedy committed:
Put fixed-key event-adds in EditorWindow.__init__ for now.

terryjreedy committed:
Don't warn if new core keys not in user key config.

terryjreedy committed:
Make code context inoperative in outwin, shell.

terryjreedy commented:
This may be complete enough, but I may decide to change the new mac bindings first, as I really do not like to leave IDLE in a can't-release regressed state after a commit.

'Code context'  is visible on the option menu for Shell and Output Window but not operative.  It should be gone (as before) or grayed out, but the current situation could be tolerated in a release.

terryjreedy committed:
Change new bindings for MacOsx.

terryjreedy commented:
I ran test suite a couple of times, but now there are two failures to look at.

FAIL: test_get_current_keyset (idlelib.idle_test.test_config.IdleConfTest)
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/idlelib/idle_test/test_config.py", line 526, in test_get_current_keyset
    self.assertNotEqual(conf.GetCurrentKeySet(), conf.GetKeySet(conf.CurrentKeys()))

FAIL: test_click_help_source (idlelib.idle_test.test_query.HelpsourceGuiTest)
Traceback (most recent call last):
  File "C:\projects\cpython\lib\idlelib\idle_test\test_query.py", line 347, in test_click_help_source
    Equal(dialog.result, ('__test__', prefix + __file__))
AssertionError: Tuples differ: ('__test__', 'C:\\projects\\cpython\\lib\\idlelib\\idle[17 chars].py') != ('__test__', 'file://C:\\projects\\cpython\\lib\\idleli[24 chars].py')
First differing element 1:
'C:\\projects\\cpython\\lib\\idlelib\\idle_test\\test_query.py'
'file://C:\\projects\\cpython\\lib\\idlelib\\idle_test\\test_query.py'

terryjreedy committed:
Disable test dependent on extension Alt keys.

terryjreedy commented:
I ran tests before changing Mac keys.  I disabled test that depended on there being Alt keys in Mac or extension bindings.  The other failure was due to sys.platform not being restored in first test.

terryjreedy committed:
Minor edits.

terryjreedy committed:
File out news blurb explaining effect for users.

terryjreedy commented:
There are some regressions in the untested behavior of Config Dialog.  I suspect they occurred before this patch, but I need to check before merging.

terryjreedy committed:
Don't compute ('') during integer input in entry box.

terryjreedy commented:
By checking out master, I verified that the problems I found yesterday pre-existed this patch.  The new one I found today required a fix to the new code.

miss-islington commented:
Thanks @wohlganger for the PR, and @terryjreedy for merging it 🌮🎉.. I'm working now to backport this PR to: 3.6.
🐍🍒⛏🤖 I'm not a witch! I'm not a witch!
